src/main/java/reposense/report/ReportGenerator.java

```diff
@@ -107,23 +115,35 @@ public static void generateReposReport(List<RepoConfiguration> configs, String o
      * Clone, analyze and generate the report for repositories in {@code repoLocationMap}.
      * Performs analysis and report generation of each repository in parallel with the cloning of the next repository.
      */
-    private static void cloneAndAnalyzeRepos(
-            Map<RepoLocation, List<RepoConfiguration>> repoLocationMap, String outputPath) throws IOException {
+    private static void cloneAndAnalyzeRepos(List<RepoConfiguration> configs, String outputPath) {
+        Map<RepoLocation, List<RepoConfiguration>> repoLocationMap = groupConfigsByRepoLocation(configs);
         RepoCloner repoCloner = new RepoCloner();
         RepoLocation clonedRepoLocation = null;
 
-        for (RepoLocation location : repoLocationMap.keySet()) {
-            repoCloner.clone(outputPath, repoLocationMap.get(location).get(0));
+        List<RepoLocation> repoLocationList = new ArrayList<>(repoLocationMap.keySet());
+        Set<RepoLocation> successfullyClonedRepos = new HashSet<>();
+
+        // clones another repo in parallel while analyzing a cloned repo (if any)
+        for (RepoLocation location : repoLocationList) {
+            repoCloner.clone(repoLocationMap.get(location).get(0));
 
             if (clonedRepoLocation != null) {
                 analyzeRepos(outputPath, repoLocationMap.get(clonedRepoLocation),
                         repoCloner.getCurrentRepoDefaultBranch());
+                successfullyClonedRepos.add(clonedRepoLocation);
             }
-            clonedRepoLocation = repoCloner.getClonedRepoLocation(outputPath);
+            clonedRepoLocation = repoCloner.getClonedRepoLocation();
         }
         if (clonedRepoLocation != null) {
-            analyzeRepos(outputPath, repoLocationMap.get(clonedRepoLocation), repoCloner.getCurrentRepoDefaultBranch());
+            analyzeRepos(outputPath, repoLocationMap.get(clonedRepoLocation),
+                    repoCloner.getCurrentRepoDefaultBranch());
+            successfullyClonedRepos.add(clonedRepoLocation);
         }
+
+        // handles configs that failed to clone and clean up the cloned repo directory.
+        handleCloningFailed(configs.stream()
+                .filter(config -> !successfullyClonedRepos.contains(config.getLocation()))
+                .collect(Collectors.toList()));
```

Comment:
Is it simpler to store a list of `RepoConfiguration` that failed analysis then pass it to `handleCloningFailed()`

Reply:
I think doing so may be complicated as well due to the nature of asynchronous cloning process.

Assuming that I maintain a list of uncloned `RepoConfiguration` by initializing it with all `configs`. Every time a `RepoLocation` has been successfully cloned, I'll need to get a list  of `RepoConfiguration` that's successfully cloned and remove from the list of uncloned `RepoConfiguration`. Then only at the end I have a final list of `RepoConfiguration` that failed to clone and pass it to `handleCloningFailed()`.

Reply:
Hmm, I was thinking that we can append failed repos to a list in the `catch` part of the code.
But, devs may forget to append when create new git exceptions.

So, your approach may be better in this case. 😃